odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,896 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+import numpy as np
+
+from odl.operator.operator import (
+    Operator, OperatorComp, OperatorLeftScalarMult, OperatorRightScalarMult,
+    OperatorRightVectorMult, OperatorSum)
+from odl.operator.default_ops import (IdentityOperator, ConstantOperator)
+from odl.solvers.advanced import (proximal_arg_scaling, proximal_translation,
+                                  proximal_quadratic_perturbation,
+                                  proximal_zero)
+
+
+__all__ = ('Functional', 'FunctionalLeftScalarMult',
+           'FunctionalRightScalarMult', 'FunctionalComp',
+           'FunctionalRightVectorMult', 'FunctionalSum', 'FunctionalScalarSum',
+           'FunctionalTranslation', 'FunctionalLinearPerturb')
+
+
+class Functional(Operator):
+
+    """Implementation of a functional class.
+
+    A functional is an operator ``f`` that maps from some domain ``X`` to the
+    field of scalars ``F`` associated with the domain:
+
+        ``f : X -> F``.
+
+    Notes
+    -----
+    The implementation of the functional class assumes that the domain
+    :math:`X` is a Hilbert space and that the field of scalars :math:`F` is a
+    is the real numbers. It is possible to create functions that do not fulfil
+    these assumptions, however some mathematical results might not be valide in
+    this case. For more information, see
+    http://odl.readthedocs.io/guide/in_depth/functional_guide.html.
+    """
+
+    def __init__(self, space, linear=False, grad_lipschitz=np.nan):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        space : `LinearSpace`
+            The domain of this functional, i.e., the set of elements to
+            which this functional can be applied.
+        linear : bool, optional
+            If `True`, the functional is considered as linear.
+        grad_lipschitz : float, optional
+            The Lipschitz constant of the gradient. Default: ``nan``
+        """
+        Operator.__init__(self, domain=space,
+                          range=space.field, linear=linear)
+        self.__grad_lipschitz = float(grad_lipschitz)
+
+    @property
+    def grad_lipschitz(self):
+        """Lipschitz constant for the gradient of the functional"""
+        return self.__grad_lipschitz
+
+    @grad_lipschitz.setter
```

Comment:
Not really any practical one, but I think that @kohr-h's comment on this earlier was valide: assume I make a composition `f(A(x))`. If `A` is linear, maybe the user can calculate what `grad_lipschitz` is and wants to manually change the value.

Also: in writing this comment I see that `FunctionalComp` sets it to `np.inf`. I need to go through and see that it is set to `np.nan` in all places where we don't know.